### PR TITLE
Fix progressBar endDuration when using `stages`

### LINKED ISSUE #347: When using `stages`, the duration and progressbar on console output is wrong
screen shot 2017-10-20 at 10 56 53

@liclac I could try a PR but perhaps better to wait for the CLI refactor.

### PR DESCRIPTION
**Pull request by ppcano**
closes #347

### CODE REVIEW

**Comment by ppcano** [comment]
`master` rebased.

**Comment by liclac** [comment]
Wrong place to fix this. Never duplicate information, that's a hack that *never* ends well. Instead change the `Right` function for the progress bar to check both what `GetEndDuration()` and `GetStages()` say, and use whichever one is valid (for `GetStages()`, `len(stages) > 0`) and lower.

**Comment by ppcano** [comment]
@liclac Thanks.

Now, I have updated only the `Right` function.  The first attempt also updated the `duration` value of the test summary. The new change does not update the `duration` ( - ), but I am not sure if this is desired. 


screen shot 2017-11-29 at 13 54 10

The intention of the PR was to clarify the duration when using `stages`; IMO, updating the `progressbar` is enough to clarify it. We could evaluate later if more changes are required.

File: cmd/run.go [hunk 1]
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
 				}
 				precision := 100 * time.Millisecond
 				atT := engine.Executor.GetTime()
-				if endT := engine.Executor.GetEndTime(); endT.Valid {
+
```
Comment (liclac) [inline cmd/run.go]: Drop this newline.

File: cmd/run.go [hunk 2]
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
...
+				endT := engine.Executor.GetEndTime()
+				if stages := engine.Executor.GetStages(); len(stages) > 0 {
```
Comment (liclac) [inline cmd/run.go]: Right, I forgot `SumStages()` returns a `NullDuration`, just make this an `endT := SumStages(...); endT.Valid`.

File: cmd/run.go [hunk 3]
```diff
@@ -268,7 +268,13 @@ a commandline interface for interacting with it.`,
...
+					endT = local.SumStages(stages)
```
Comment (liclac) [inline cmd/run.go]: If this is gonna be used outside of `local`, move it to `lib`.

**Comment by liclac** [comment]
Just update the progress bar with the result of `SumStages()` then.

https://github.com/ppcano/k6/blob/926bb0089d71d82cd5ed398ed6e0b0ca34134842/cmd/run.go#L322

**Comment by ppcano** [comment]
@liclac I have added your changes.

I am not sure how to implement if/else in a more idiomatic way:

```
endT := lib.SumStages(engine.Executor.GetStages())
if !endT.Valid {
   endT = engine.Executor.GetEndTime()
}
if endT.Valid {
     prog = float64(engine.Executor.GetTime()) / float64(endT.Duration)
}

```
```
endT := lib.SumStages(engine.Executor.GetStages())
if !endT.Valid {
    endT = engine.Executor.GetEndTime()
}
if endT.Valid {
   ...					
}
```

Feel free to suggest. Thanks.

**Comment by liclac** [comment]
That's fine.

But you also just switched the priority (stages > end). The correct behaviour would be to get both times, then use whichever is valid and lower.

```go
endT := GetEndDuration()
stageEndT := SumStages()
if !endT.Valid || endT.Duration < stageEndT.Duration {
  endT = stageEndT
}
```

**Comment by ppcano** [comment]
@liclac 

> The correct behaviour would be to get both times, then use whichever is valid and lower.

Done at 5785f20cd465e850bdc984d0f1f0927bc51a87df

```
if !endT.Valid || endT.Duration > stageEndT.Duration {
  endT = stageEndT
}
```

Let me know if you want me to `squash` all the commits or there is something else to include.

File: cmd/run.go [hunk 4]
```diff
@@ -273,7 +273,13 @@ a commandline interface for interacting with it.`,
 				}
 				precision := 100 * time.Millisecond
 				atT := engine.Executor.GetTime()
-				if endT := engine.Executor.GetEndTime(); endT.Valid {
+				stagesEndT := lib.SumStages(engine.Executor.GetStages())
+				endT := engine.Executor.GetEndTime()
+				//get both times and use whichever is valid and lower
```
Comment (liclac) [inline cmd/run.go]: This comment is kind of unnecessary. If you feel it needs to be here, capitalise the G and put a space after the "//".

**Comment by ppcano** [comment]
@liclac Removed the comment. Done, Thanks.

**Pull request merged**